Add service for entity registrant detection

dhaval-parekh opened the pull request:
## Summary

<!-- Please reference the issue(s) this PR fixes, if one exists. For significant changes, opening an issue first for discussion is usually preferable. -->
Fixes #6516 

This PR adds new services "EntityRegistrantDetection", which detect the function which registers new entities such as post type, taxonomy, shortcode or block. And provide that info via REST endpoint.

**Sample REST response.**
```json
{
	"post_types":{
		"product":{
			"name":"Products",
			"slug":"product",
			"description":"This is where you can browse products in this store.",
			"hierarchical":false,
			"source":{
				"type":"plugin",
				"name":"woocommerce",
				"file":"includes\/class-wc-post-types.php",
				"line":286,
				"function":"WC_Post_Types::register_post_types",
				"hook":"init",
				"priority":5
			}
		},
	},
	"taxonomies":{
		"product_type":{
			"name":"Product type",
			"slug":"product_type",
			"description":"",
			"hierarchical":false,
			"source":{
				"type":"plugin",
				"name":"woocommerce",
				"file":"includes\/class-wc-post-types.php",
				"line":37,
				"function":"WC_Post_Types::register_taxonomies",
				"hook":"init",
				"priority":5
			}
		},
	},
	"blocks":{
		"woocommerce\/product-summary":{
			"name":"woocommerce\/product-summary",
			"title":"woocommerce\/product-summary",
			"description":"",
			"category":null,
			"attributes":[
				
			],
			"is_dynamic":true,
			"source":{
				"type":"plugin",
				"name":"woocommerce",
				"file":"packages\/woocommerce-blocks\/src\/BlockTypesController.php",
				"line":57,
				"function":"Automattic\\WooCommerce\\Blocks\\BlockTypesController::register_blocks",
				"hook":"init",
				"priority":10
			}
		},
	},
	"shortcodes":{
		"product":{
			"tag":"product",
			"source":{
				"type":"plugin",
				"name":"woocommerce",
				"file":"includes\/class-wc-shortcodes.php",
				"line":19,
				"function":"WC_Shortcodes::init",
				"hook":"init",
				"priority":10
			}
		},
	}
}
```

## Checklist

- [ ] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [ ] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

dhaval-parekh committed:
Add service for entity registrant detection

dhaval-parekh committed:
Update .phpstorm.meta.php and Fix typos

dhaval-parekh committed:
Implement ArrayAccess in CallbackWrapper class

dhaval-parekh committed:
Add test cases for EntityRegistrantDetectionManager class

dhaval-parekh committed:
Add test cases for  REST controller of EntityRegistrantDetectation service

dhaval-parekh committed:
Add test cases for wrapper callback of EntityRegistrantDetectation service

dhaval-parekh committed:
Prevent to detect source of default entities

westonruter commented on src/EntityRegistrantDetection/RestController.php:
```diff
@@ -0,0 +1,132 @@
+
+		return rest_ensure_response(
+			[
+				'post_types' => EntityRegistrantDetectionManager::$post_types_source,
```
There shouldn't need to be static access for `EntityRegistrantDetectionManager`. Instead, `EntityRegistrantDetectionManager` should be added as a dependency in the constructor of this class, and it will automatically get constructed so you can then do `$this->detection_manager->post_types_source`, for example.

westonruter commented on src/EntityRegistrantDetection/CallbackWrapper.php:
```diff
@@ -0,0 +1,347 @@
+
+			$this->registered_entities[ $entity_type ] = $different;
+
+			EntityRegistrantDetectionManager::add_source( $entity_type, $different, $this->source );
```
We should avoid static method usage.

westonruter replied:
I realize that complicates the architecture, but maybe not so much. If `EntityRegistrantDetectionManager` is simply added among the services returned by `\AmpProject\AmpWP\AmpWpPlugin::get_shared_instances()`, then only one copy of `EntityRegistrantDetectionManager` will be instantiated and this will then have the same effect.

westonruter commented on src/EntityRegistrantDetection/CallbackWrapper.php:
```diff
@@ -0,0 +1,347 @@
+	 *     @type array $shortcode List of default shortcode.
+	 * ]
+	 */
+	const DEFAULT_ENTITIES = [
```
Instead of storing this here, I think it would be better to have a `$initial_entities` variable which is `null` to begin with, then when the callback wrapper is invoked it can check if `$this->initial_entities === null`, and of so, populate the initial array with what is registered. 

Nevertheless, I wonder if it is necessary? If we capture all of the registered entities at `prepare` and then diff them with what is registered at `finalize`, why do we need to also need the default entities?

dhaval-parekh replied:
> Instead of storing this here, I think it would be better to have a $initial_entities variable which is null to begin with, then when the callback wrapper is invoked it can check if $this->initial_entities === null, and of so, populate the initial array with what is registered.

Currently, We are doing the same but with `$registered_entities`. In `prepare()`, we collect all registered entities and in `finalize()` we update the same property with a difference of registered entities.

> Nevertheless, I wonder if it is necessary? If we capture all of the registered entities at prepare and then diff them with what is registered at `finalize()`, why do we need to also need the default entities?

Since We are checking all the callbacks from all hooks. It will also track the WordPress default functions which registers default entities. which I think is unnecessary to track. (Since the idea to detect if any theme/plugin is registering entities or not) and will increase the REST API payload.
So to exclude WordPress default entities in `finalize()` this const is being used.

westonruter replied:
Sorry I haven't been able to respond to this yet. I'll have to dig more deeply into the PR and its changes in order to be able to reply. That'll have to come next week.

westonruter commented on src/EntityRegistrantDetection/CallbackWrapper.php:
```diff
@@ -0,0 +1,347 @@
+	 */
+	protected function set_source() {
+
+		$callback_reflection = Services::get( 'dev_tools.callback_reflection' );
```
This should obtain the service via dependency injection.

westonruter commented on src/EntityRegistrantDetection/CallbackWrapper.php:
```diff
@@ -0,0 +1,347 @@
+	 *     @type string   $hook
+	 * ]
+	 */
+	public function __construct( $callback ) {
```
This can add service dependencies, including `EntityRegistrantDetectionManager` and `CallbackReflection`.

westonruter commented on src/EntityRegistrantDetection/EntityRegistrantDetectionManager.php:
```diff
@@ -0,0 +1,270 @@
+	 */
+	protected static function wrapped_callback( $callback ) {
+
+		return new CallbackWrapper( $callback );
```
This could then do:

```suggestion
		$wrapper = $this->injector->make( CallbackWrapper::class );
		$wrapper->set_callback( $callback );
		return $wrapper;
```

Maybe `$callback` could still be supplied among the `$arguments` of the `make` method too.

westonruter commented on src/EntityRegistrantDetection/RestController.php:
```diff
@@ -0,0 +1,132 @@
+			);
+		}
+
+		return current_user_can( 'manage_options' );
```
Let's move the `manage_options` check up to the top of this method.

milindmore22 commented:
PHP Warning:  hash_equals(): Expected known_string to be a string, null given in /srv/www/amp-local/public_html/wp-content/plugins/amp/src/EntityRegistrantDetection/EntityRegistrantDetectionManager.php on line 109

dhaval-parekh committed:
Remove static method to add source.
Use injector instead of direct initilizing object
Add dependancy in counstruct method instead of direct initilizing object

dhaval-parekh commented:
> PHP Warning: hash_equals(): Expected known_string to be a string, null given in /srv/www/amp-local/public_html/wp-content/plugins/amp/src/EntityRegistrantDetection/EntityRegistrantDetectionManager.php on line 109

I have addressed this in [0e5aa72](https://github.com/ampproject/amp-wp/pull/6682/commits/0e5aa72ed1571b6d8e14f944224628d519673c39) commit.

dhaval-parekh committed:
Fix unit test cases

dhaval-parekh committed:
Fix unit test cases

dhaval-parekh committed:
Update unit test cases for WordPress prior to 5.0

dhaval-parekh committed:
Update unit test cases for WordPress prior to 5.0

dhaval-parekh committed:
Update unit test case

dhaval-parekh committed:
Fix undefined index in unit test cases

dhaval-parekh committed:
Fix undefined index in unit test cases

schlessera commented on src/EntityRegistrantDetection/CallbackWrapper.php:
```diff
@@ -0,0 +1,364 @@
+
+		$this->prepare();
+
+		$result = call_user_func_array(
+			$this->get_callback_function(),
+			array_slice( $args, 0, (int) $this->callback['accepted_args'] )
+		);
```
Is there a reason why the outdated and slow `call_user_func_array()` is used here?
This could be rewritten to use a faster variadic arguments call:
```suggestion
		$callback = $this->get_callback_function();
		$result   = $callback( ... array_slice( $args, 0, (int) $this->callback['accepted_args'] ));
```
Note that it cannot be done as a one-liner because of PHP 5.6 parsing.

schlessera commented on src/EntityRegistrantDetection/EntityRegistrantDetectionManager.php:
```diff
@@ -0,0 +1,305 @@
+	 */
+	protected function wrapped_callback( $callback ) {
+
+		return $this->injector->make( CallbackWrapper::class, compact( 'callback' ) );
```
Injecting the injector should always be avoided whenever possible. There are two viable options for doing so here:

**If `CallbackWrapper` is considered a service or similar:**

Instead of injecting the injector (and hence, giving control to the object of the entire rest of the application), it would be preferable to inject a `CallbackWrapperFactory` instead which is specialized on directly instantiating instances of type `CallbackWrapper`.

Using a factory is always a valid way of decoupling a `new` call, as the injector can override the factory implementation if needed.

**If `CallbackWrapper` is considered a Value Object:**

If we consider `CallbackWrapper` to be a [value object](https://en.wikipedia.org/wiki/Value_object) conceptually (which is a bit far-fetched but not entirely wrong), then it would mean we can freely `new` that value object directly, instead of having it be instantiated via the injector.

dhaval-parekh committed:
Update locgic to call callable function
